core: new inference system for constraints

alexarice opened the pull request:
I've gone through a few different refactors this week and have eventually settled on this slightly smaller refactor for the constraint system. The key addition is that each constraint now returns a dictionary of "resolvers" (name can maybe be improved) which resolve a variable from the given attribute. This allows inference to be decoupled entirely from verification, and inference should no longer be quite as fragile/maybe slightly quicker.

In this version, the resolvers are not used for verification, as I had originally planned, and instead the previous verification framework is left as is. I believe this change can enable the following changes in the future:
- Allow constraints on properties to resolve variables in a clean way (see #3318)
- Allow the introduction of "length/int constraints" for managing the length of variadics (see #3416)
- Potentially facilitate extensions to assembly format (i.e. directly supplying a variable)
- Resolvers could be used to automate construction of initialiser methods for operation. 

PS. as of time of writing it appears this saves 4 lines of code

superlopuh commented on xdsl/dialects/bufferization.py:
```diff
@@ -41,42 +48,29 @@ class TensorMemrefInferenceConstraint(VarConstraint[Attribute]):
-            raise VerifyException(
-                f"Unexpected {self.name} - attribute must be ContainerType"
-            )
+        if isa(attr, TensorType[Attribute]):
```
```suggestion
        if isinstance(attr, TensorType):
```

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -1,17 +1,29 @@
+ResolveType: TypeAlias = Attribute | Sequence[Attribute]
+
+
+def print_resolve_type(r: ResolveType | None) -> str:
+    if isinstance(r, Attribute):
+        return str(r)
+    elif isinstance(r, Sequence):
+        return str(tuple(str(x) for x in r))
+    return "None"
```
```suggestion
def _format_resolve_type(r: ResolveType | None) -> str:
    if isinstance(r, Attribute):
        return str(r)
    elif isinstance(r, Sequence):
        return str(tuple(str(x) for x in r))
    return "None"
```

superlopuh replied:
Also a doc string would be good

alexarice replied:
Seems like this is dead code so I've removed it

superlopuh commented on tests/irdl/test_declarative_assembly_format.py:
```diff
@@ -1748,10 +1748,12 @@ class OneOperandOneResultNestedOp(IRDLOperation):
-        ParseError,
-        match="Verification error while inferring operation type: ",
+        VerifyException,
+        match="i32 should be of base attribute test.param_one",
```
This seems like it conveys the message a bit more directly
```suggestion
        match="unexpected attribute i32, expected test.param_one",
```

superlopuh replied:
hmm actually is this just an existing constraint string? In that case I won't recommend PR creep, it might just be worth changing in the future.

alexarice replied:
The key change here is that the exception happens during verification, rather than at parse time

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -560,22 +571,24 @@ def verify(
+        variables: dict[str, ResolveType],
+    ) -> Sequence[AttributeCovT]:
+        v = variables[self.name]
+        return cast(Sequence[AttributeCovT], v)
```
Does this work?
```suggestion
        assert not isinstance(V, Attribute)
        return v
```

alexarice replied:
No

superlopuh replied:
Sad

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -591,19 +604,19 @@ def verify(
+        variables: dict[str, ResolveType],
+    ) -> Sequence[AttributeCovT]:
+        attr = self.constr.infer(variables)
+        return tuple(attr for _ in range(length))
```
```suggestion
        return (attr,) * length
```

superlopuh reviewed (commented):
Overall looks good, although I'm still not sure what the motivation for this is. Is it better error messages? Or is there more downstream? Also, my first instinct would be to somehow be a bit more lazy about the resolver situation, would it make sense to make them into lazy model objects, that let the user iterate over the keys, and define each function only once, instead of creating wrappers of wrappers. Not 100% sure if it would work, but might be worth a try in this PR?

alexarice commented on xdsl/irdl/constraints.py:
```diff
@@ -1,17 +1,29 @@
 from xdsl.utils.exceptions import VerifyException
 from xdsl.utils.runtime_final import is_runtime_final
 
+_AttributeCovT = TypeVar("_AttributeCovT", bound=Attribute, covariant=True)
```
This is (possibly accidentally) moved instead of being new. It's used in the definition of `__or__` (which uses both this and the existing one.

superlopuh replied:
Ah ok sounds good, thank you

alexarice replied:
I've moved it back

alexarice commented:
I tried to put some motivation in the PR description and there is a bit more discussion in #3318 about needing to verify properties during parsing with the old system (I have rebased that PR onto this PR).

superlopuh commented:
Thank you, happy to approve once the CI passes

alexarice commented on xdsl/irdl/constraints.py:
```diff
@@ -72,24 +74,23 @@ def verify(
         ...
 
-    def get_resolved_variables(self) -> set[str]:
+    def get_resolvers(self) -> dict[str, Callable[[AttributeCovT], ResolveType]]:
```
It might be worth making a type alias for this, as it appears in a few places

alexarice replied:
Replaced this by object-based resolvers so this is no longer relevant

alexarice commented:
>  Also, my first instinct would be to somehow be a bit more lazy about the resolver situation, would it make sense to make them into lazy model objects, that let the user iterate over the keys, and define each function only once, instead of creating wrappers of wrappers. Not 100% sure if it would work, but might be worth a try in this PR?

I'm also not sure how happy I am about stacks of wrappers. Another option which could be worth trying is to have a resolver be a list of "accessors" where each accessor is a piece of data (such as the attribute name, or operand index). I haven't tried putting this into action so I'm not sure how messy it would get.

Could you expand on what you mean by "lazy model objects"?

superlopuh commented:
Something like
```
class AbstractResolver(abc.ABC):
    @abstractmethod
    def iter_variables() -> Iterator[str]: ...

     @abstractmethod
    def resolve(attr: Attribute) -> Attribute | Sequence[Attribute]:
        ...

@dataclass
class WhateverResolver(AbstractResolver):

    a: AbstractResolver
    b: AbstractResolver

    def iter_variables() -> Iterator[str]:
        yield from a.iter_variables()
        yield from b.iter_variables()

    def resolve(attribute: Attribute) -> Attribute | Sequence[Attribute]:
        return a.resolve(attribute) if creative_condition else b.resolve(attribute)
```

alexarice commented:
I like that idea, I'll give it a go next week

alexarice commented:
Have merged #3464 here, so this is ready for review again (Think there are no outstanding open comments)

superlopuh commented on xdsl/irdl/declarative_assembly_format_parser.py:
```diff
@@ -241,20 +243,46 @@ def add_reserved_attrs_to_directive(self, elements: list[FormatDirective]):
+        resolvers = dict[str, Resolver[ParsingState]]()
         for i, (_, operand_def) in enumerate(self.op_def.operands):
             if self.seen_operand_types[i]:
-                seen_variables |= operand_def.constr.get_resolved_variables()
+                resolvers.update(
+                    {
+                        v: self._OperandResultResolver(i, True, r)
+                        for v, r in operand_def.constr.get_resolvers().items()
+                    }
+                )
         for i, (_, result_def) in enumerate(self.op_def.results):
             if self.seen_result_types[i]:
-                seen_variables |= result_def.constr.get_resolved_variables()
-        return seen_variables
+                resolvers.update(
+                    {
+                        v: self._OperandResultResolver(i, False, r)
+                        for v, r in result_def.constr.get_resolvers().items()
+                    }
+                )
+        return resolvers
```
What happens if there's an overlap in `v`s? Worth a comment if that's not possible? My instinct here would be to create two dictionaries, probably with two dictionary comprehensions, and then take their union. (Maybe assert that the intersection of keys is empty)

alexarice replied:
If there's an overlap in `v`s then the second one encountered will be used. This would be happen if a variable can be inferred from two different places in the assembly format. Which resolver is used shouldn't make any difference for any operation that successfully verifies.

superlopuh replied:
how do you feel about a resolver that asserts at runtime that the two resolved results are the same?

alexarice replied:
I imagine this would be easy to do with the new class-based system. I'm not sure how I feel about putting what is effectively verification as a check in parsing.

superlopuh commented on xdsl/dialects/bufferization.py:
```diff
@@ -42,42 +47,29 @@ class TensorMemrefInferenceConstraint(VarConstraint[Attribute]):
-            raise VerifyException(
-                f"Unexpected {self.name} - attribute must be ContainerType"
-            )
+        if isa(attr, TensorType[Attribute]):
+            memref_type = MemRefType(attr.element_type, attr.shape)
+            return self.memref_var_constr.verify(memref_type, constraint_context)
+        if isa(attr, UnrankedTensorType[Attribute]):
+            memref_type = UnrankedMemrefType.from_type(attr.element_type)
+
+        raise VerifyException(f"Expected TensorType or UnrankedTensorType, got {attr}")
```
the logic here seems to be changed

alexarice replied:
It is, though it appears I forgot to amend the doc comment. In my opinion the previous constraint was an abuse of the particular workings of the previous system. The one I have written here is a bit more well-behaved with respect to inference, and can be used to achieve the same effect as the old one. I'll update the doc comment.

superlopuh replied:
Would it make sense to merge this change as a separate PR, while we discuss the details of this one?

alexarice replied:
Sure, I'll get that ready first thing tomorrow

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -72,24 +82,23 @@ def verify(
+        return dict()
 
-    def can_infer(self, constraint_names: set[str]) -> bool:
+    def can_infer(self, variables: Set[str]) -> bool:
```
How do you feel about this?
```suggestion
    def can_infer(self, `var_constraint_names`: Set[str]) -> bool:
```
I probably should have commented this on the previous PR, my bad

alexarice replied:
I assume you didn't mean to include the backticks? It seems reasonable to me.

superlopuh replied:
I didn't indeed

alexarice replied:
There was a small chance it was some obscure python syntax I'd not seen before

superlopuh replied:
it's obscure Swift and Mojo syntax!

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -339,21 +351,19 @@ def verify(
+    def get_resolvers(self) -> dict[str, Resolver[AttributeCovT]]:
+        d: dict[str, Resolver[AttributeCovT]] = dict()
+        for constr in self.attr_constrs:
+            d |= constr.get_resolvers()
```
what happens if there are overlapping keys here?

alexarice replied:
Overlapping keys get overwritten

superlopuh replied:
Is the idea here also that the expectation is that the resolvers will return identical results, so this is not an issue?

alexarice replied:
Yes

superlopuh commented on xdsl/dialects/builtin.py:
```diff
@@ -1637,6 +1632,49 @@ def constr(
+class TensorOrMemrefOf(
+    GenericAttrConstraint[TensorType[AttributeCovT] | MemRefType[AttributeCovT]]
+):
+    """A type constraint that can be nested once in a vector or a tensor."""
```
```suggestion
    """A type constraint that can be nested once in a memref or a tensor."""
```

superlopuh commented:
How do you feel about "variable assignment" as opposed to "resolving" variables?

alexarice commented:
> How do you feel about "variable assignment" as opposed to "resolving" variables?

What would the concrete change be here? Open to different names for some/any of these things, I just went with the first thing that came to me and stuck with it

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -55,6 +56,15 @@ def update(self, other: ConstraintContext):
+_T = TypeVar("_T")
+
+
+class Resolver(Generic[_T], abc.ABC):
+    @abstractmethod
+    def resolve(self, a: _T) -> ResolveType: ...
```
This would benefit from a lot more doc commenting, seems like the perfect place for a long-form explanation of why this is useful and how the whole mechanism works. I'm hoping that a plain English explanation will help pick the right names for the functions and function arguments involved

math-fehr commented on xdsl/irdl/constraints.py:
```diff
@@ -55,6 +56,15 @@ def update(self, other: ConstraintContext):
 
 _AttributeCovT = TypeVar("_AttributeCovT", bound=Attribute, covariant=True)
 
+ResolveType: TypeAlias = Attribute | Sequence[Attribute]
```
Can you add documentation to that alias, so it makes sense what it is?

math-fehr commented on xdsl/irdl/constraints.py:
```diff
@@ -55,6 +56,15 @@ def update(self, other: ConstraintContext):
+_T = TypeVar("_T")
+
+
+class Resolver(Generic[_T], abc.ABC):
```
Isn't this an `AttrInference` or something?
I feel using the `Inference` word makes more sense than `Resolve` here

alexarice replied:
I was considering something along the lines of `Extractor`. The idea is that it describes how to get the value of a variable from the input `_T` (which could be an attribute, sequence of attributes, ParsingState etc)

math-fehr replied:
Hmm, I see what you mean, I think `Extractor` makes more sense then indeed! Or `VarExtractor` to be more precise?

math-fehr commented on xdsl/irdl/constraints.py:
```diff
@@ -1,14 +1,15 @@
+import abc
 from abc import ABC, abstractmethod
-from collections.abc import Sequence
+from collections.abc import Sequence, Set
```
Is there a reason you are using `Set` instead of `set`?
AFAIK, we should always prefer `set` since Python 3.something?

alexarice replied:
`Set` is more general and in particular includes `dict.keys()`. Note that `collections.abc.Set =/= typing.Set`

math-fehr replied:
Oh god, I did not know about `typing.Set` versus `collections.abc.Set`...

alexarice replied:
Makes autocompletion very annoying, especially since `typing.Set` is useless since Python 3.something as you said

superlopuh replied:
Also Set and Dict are immutable collections if I remember correctly, it also helps with variance things

math-fehr commented on xdsl/irdl/constraints.py:
```diff
@@ -72,24 +82,23 @@ def verify(
         Check if there is enough information to infer the attribute given the
         constraint variables that are already set.
         """
         # By default, we cannot infer anything.
         return False
```
Because we have get_resolvers, this function should always be defined?
Is there a reason we could infer something that doesn't have a resolver, or vice versa?

alexarice replied:
Any `EqAttrConstraint` can be inferred but does not resolve any variables. A `ParamAttrConstraint` may be able to provide variable resolutions without being totally inferrable

alexarice replied:
And in general there are lots of things that cannot be inferred, and must be given in the assembly format

math-fehr replied:
Oh yes you're right, my bad.

math-fehr reviewed (approved):
Looks neat!
I'm trusting Sasha for the review, but on my end that looks good!

alexarice commented:
I believe all comments have been resolved. 
- I've renamed `Resolvers` to `VarExtractors` and given them a (admittedly small) docstring. 
- I have also introduced a `MergeExtractor` which is used to record that two extractors have been provided for the same variable. This performs a runtime check to determine that each extractor would return the same thing. I have added 3 tests for each case where this is used, using slightly contrived operations.
-  I have rebased on top of the change to the constraint in bufferization.

alexarice commented:
Forgot to change the `can_infer` method argument name, will do that now

superlopuh commented on tests/dialects/test_bufferization.py:
```diff
@@ -40,13 +39,13 @@ def test_tensor_from_memref_inference():
     )
     assert constr2.can_infer(set())
-    assert constr2.infer(ConstraintContext()) == TensorType(f64, [10, 20, 30])
+    assert constr2.infer(dict()) == TensorType(f64, [10, 20, 30])
```
?
```suggestion
    assert constr2.infer({}) == TensorType(f64, [10, 20, 30])
```

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -55,6 +56,73 @@ def update(self, other: ConstraintContext):
+            )
+
+    def flatten(self) -> Iterator[VarExtractor[_T]]:
+        return self.extractors.__iter__()
```
```suggestion
        return iter(self.extractors)
```

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -470,23 +557,22 @@ def verify(
-    """
-    Constrain a range of attributes to a certain value.
-    """
+    """Constrain an range of attributes to certain values."""
```
```suggestion
    """Constrain a range of attributes to certain values."""
```

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -500,32 +586,39 @@ def verify(
+        the value of the attribute range which is verified by this constraint.
         """
-        return set()
+        return dict()
```
```suggestion
        return {}
```

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -72,24 +140,25 @@ def verify(
+        the value of the attribute which is verified by this constraint.
         """
-        return set()
+        return dict()
```
```suggestion
        return {}
```

alexarice commented:
@superlopuh do you feel the extractor merging is worth the added complexity? My examples in the tests feel a bit contrived

superlopuh commented:
I feel like it's OK, I'm just imagining the headache that would emerge from trying to debug this, and it all seems worth it.

superlopuh commented:
I just merged the small conflicting PR, hopefully will be easy to resolve. I think the best order now is for you to merge this and for me to resolve my draft PRs.
